File: src/host/lib/response.ml
(**********************************************************************)
(*                                                                    *)
(*              This file is part of the RFSM package                 *)
(*                                                                    *)
(*  Copyright (c) 2018-present, Jocelyn SEROT.  All rights reserved.  *)
(*                                                                    *)
(*  This source code is licensed under the license found in the       *)
(*  LICENSE file in the root directory of this source tree.           *)
(*                                                                    *)
(**********************************************************************)

type t = 
  | Version of string
  | CompilationOk of string list (* list_of_generated_files *)
  | CompilationFailed of string (* error message *)
  | CheckingOk of string list * string list (* read vars, written vars *)
  | CheckingFailed of string (* error_message *)
  | Error of string
  | None
  [@@deriving show]

let string_of_json j = 
  let no_nl c = if c = '\n' then ' ' else c in
  j |> Yojson.Basic.pretty_to_string  |> String.map no_nl

let to_json (r: t) : Yojson.Basic.t =
  match r with
  | Version s ->
      `Assoc [("kind", `String "version"); ("version", `String s)]
  | CompilationOk files ->
      `Assoc [
        ("kind", `String "compiled");
        ("result", `Bool true);
        ("files", `List (List.map (fun s -> `String s) files))
      ]
  | CompilationFailed msg ->
      `Assoc [
        ("kind", `String "compiled");
        ("result", `Bool false);
        ("message", `String msg)
      ]
  | CheckingOk (rds,wrs) ->
      `Assoc [
        ("kind", `String "checked");
        ("result", `Bool true);
        ("rds", `List (List.map (fun v -> `String v) rds));
        ("wrs", `List (List.map (fun v -> `String v) wrs));
      ]
  | CheckingFailed msg ->
      `Assoc [
        ("kind", `String "checked");
        ("result", `Bool false);
        ("message", `String msg)
      ]
  | Error msg ->
      `Assoc [("kind", `String "error"); ("message", `String msg)]
  | None ->
      `Assoc [("kind", `String "none")]

let to_string r =
  Yojson.Basic.pretty_to_string (to_json r)

(* let to_string r = 
 *   let j = match r with
 *   | Version s ->
 *     `Assoc [("version", `String s)]
 *   | Compiled (true,fs) ->
 *     `Assoc [("result",`String "ok"); ("files", `List (List.map (fun s -> `String s) fs))]
 *   | Compiled (false,[msg]) ->
 *     `Assoc [("result",`String "failed"); ("error",`String msg)]
 *   | Compiled (false,_) ->
 *     `Assoc [("error", `String "Illegal response")] (\* should not happen *\)
 *   | Checked (true,_) -> (\* TBC *\)
 *     `Assoc [("result",`String "ok")]
 *   | Checked (false,msg) ->
 *     `Assoc [("result",`String "failed"); ("error",`String msg)]
 *   | None ->
 *     `String "<none>" (\* Should not happen *\)
 *   | Error s ->
 *     `Assoc [("error", `String s)]
 *   in
 *   string_of_json j *)

exception Invalid of string
    
let of_json (json : Yojson.Basic.t) : t =
  let open Yojson.Basic.Util in
  match json |> member "kind" |> to_string with
  | "version" ->
      Version (json |> member "version" |> to_string)
  | "compiled" ->
      let res = json |> member "result" |> to_bool in
      if res then
        let files = json |> member "files" |> to_list |> List.map to_string in
        CompilationOk files
      else
        let msg = json |> member "message" |> to_string in
        CompilationFailed msg
  | "checked" ->
      let res = json |> member "result" |> to_bool in
      if res then
        let rds = json |> member "rds" |> to_list |> List.map to_string in
        let wrs = json |> member "wrs" |> to_list |> List.map to_string in
        CheckingOk (rds,wrs)
      else
        let msg = json |> member "message" |> to_string in
        CheckingFailed msg
  | "error" ->
      Error (json |> member "message" |> to_string)
  | "none" ->
      None
  | other ->
      raise (Invalid other)

File: src/guests/std/tests/server/client/response.cpp
/***********************************************************************/
/*                                                                     */
/*       This file is part of the Grasp software package               */
/*                                                                     */
/*  Copyright (c) 2019-present, Jocelyn SEROT (jocelyn.serot@uca.fr)   */
/*                       All rights reserved.                          */
/*                                                                     */
/*    This source code is licensed under the license found in the      */
/*      LICENSE file in the root directory of this source tree.        */
/*                                                                     */
/***********************************************************************/

#include "response.h"

// Public ctors
Response Response::Version(const QString &v) { return Response(Kind::Version,v); }
Response Response::Error(const QString &msg) { return Response(Kind::Error, msg); }
Response Response::CompilationFailed(const QString &msg) { return Response(Kind::Compiled, msg); }
Response Response::CheckingFailed(const QString &msg) { return Response(Kind::Checked, msg); }
Response Response::CompilationOk(const QStringList &files) { return Response(Kind::Compiled, files); }
Response Response::CheckingOk() { return Response(Kind::Checked); }
Response Response::None() { return Response(Kind::None); }

// Private ctors
Response::Response(Kind kind) : m_kind(kind) // None, CheckingOk
{
  switch ( kind ) {
  case Kind::Checked: m_success = true; break;
  default: break;
  }
}

Response::Response(Kind kind, const QString &s) : m_kind(kind) // Version, CompilationFailed, CheckingFailed, Error
{
  switch ( kind ) {
  case Kind::Version: m_version = s; break;
  case Kind::Compiled: m_success = false; m_message = s; break;
  case Kind::Checked: m_success = false; m_message = s; break;
  case Kind::Error: m_message = s; break;
  default: break; // Should not happen
  }
}

Response::Response(Kind kind, const QStringList &ss)
  : m_kind(kind), m_success(true), m_files(ss) {} // CompilationOk

Response::Kind Response::kind() const { return m_kind; }
QString Response::version() const { return m_version; }
bool Response::success() const { return m_success; }
QStringList Response::files() const { return m_files; }
QString Response::message() const { return m_message; }
QString Response::error() const { return m_error; }

QJsonObject Response::toJson() const {
    QJsonObject obj;
    switch (m_kind) {
    case Kind::Version:
        obj["kind"] = "version";
        obj["version"] = m_version;
        break;
    case Kind::Compiled: {
        obj["kind"] = "compiled";
        obj["success"] = m_success;
        QJsonArray arr;
        for (const auto &f : m_files) arr.append(f);
        obj["files"] = arr;
        break;
    }
    case Kind::Checked:
        obj["kind"] = "Checked";
        obj["success"] = m_success;
        obj["message"] = m_message;
        break;
    case Kind::Error:
        obj["kind"] = "Error";
        obj["error"] = m_error;
        break;
    case Kind::None:
        obj["kind"] = "None";
        break;
    }
    return obj;
}

Response Response::fromJson(const QJsonObject &obj) {
    QString kind = obj["kind"].toString();
    if (kind == "version") {
        return Version(obj["version"].toString());
    } else if (kind == "compiled") {
        bool success = obj["result"].toBool();
        if ( success ) {
          QStringList files;
          for (const auto &v : obj["files"].toArray()) files.append(v.toString());
          return CompilationOk(files);
          }
        else {
          QString msg = obj["message"].toString();
          return CompilationFailed(msg);
          }
    } else if (kind == "checked") {
        bool success = obj["success"].toBool();
        if ( success ) 
          return CheckingOk();
        else {
          QString msg = obj["message"].toString();
          return CheckingFailed(msg);
          }
    } else if (kind == "error") {
        return Error(obj["error"].toString());
    } else {
        return None();
    }
}

Response Response::fromString(const QString &jsonStr)
{
    QJsonParseError err;
    QJsonDocument doc = QJsonDocument::fromJson(jsonStr.toUtf8(), &err);
    if (err.error != QJsonParseError::NoError) {
        return Response::None(); // default fallback
    }
    return fromJson(doc.object());
}

